Generic Pointer to Field (#2708)
rust-lang/rfcs: text/0000-ptr-to-field.md

# Motivation
[motivation]: #motivation

The motivation for this feature is to allow safe projection through smart pointers, for example `Pin<&mut T>` to `Pin<&mut Field>`. This is a much needed feature to make `Pin<P>` more usable in safe-contexts, without the need to use unsafe to map to a field. This also can allow projection through other smart pointers like `Rc<T>`, `Arc<T>`. This feature cannot be implemented as a library effectively because it depends on the layouts of types, so it requires integration with the Rust compiler until Rust gets a stable layout (which may never happen).

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

First the core traits, type, and functions that need to be added

```rust
/// Contains metadata about how to get to the field from the parent using raw pointers
/// This is an opaque type that never needs to be stablized, and it only an implementation detail
struct FieldDescriptor {
    ...
}

/// The compiler should prevent user implementations for `Field`,
/// i.e. only the compiler is allowed to implement `Field`
/// This is to prevent people from creating fake "fields"

> trait Field {

Reviewer: This should be `unsafe trait` to represent that unsafe code can rely on its values.

Reviewer 2: Ojh, yes, thanks for the catch!

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

First the core traits, type, and functions that need to be added

```rust
/// Contains metadata about how to get to the field from the parent using raw pointers
/// This is an opaque type that never needs to be stablized, and it only an implementation detail
struct FieldDescriptor {
    ...
}

/// The compiler should prevent user implementations for `Field`,
/// i.e. only the compiler is allowed to implement `Field`
/// This is to prevent people from creating fake "fields"
unsafe trait Field {
    /// The type that the field is a part of
    type Parent;
    /// The type of the field

>     type Type;

Reviewer 3: The `Parent` could be unsized.

```rust
struct S<T: ?Sized> {
    a: u16,
    b: T,
}
let field_1 = S<dyn Debug>.a;
```

and could the `Type` be unsized?

```rust
struct S<T: ?Sized> {
    a: u16,
    b: T,
}
let field_2 = S<dyn Debug>.b;
```

Reviewer 2: Yay, more things I missed in translation from the internals! Thank you for reporting this!

    /// The metadata required to get to the field using raw pointers
    const FIELD_DESCRIPTOR: FieldDescriptor;
}

trait Project<F: Field> {
    /// The projected version of Self
    type Projection;

    fn project(self, field: F) -> Self::Projection;
}

impl<T: ?Sized> *const T {
    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type {
        // make the field pointer, this code is allowed to assume that
        // self points to a valid instance of T
        ... 
    }
}

impl<T: ?Sized> *mut T {
    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type {
        // make the field pointer, this code is allowed to assume that
        // self points to a valid instance of T
        ...
    }
}
```

Now we need some syntax to refer to the fields types. Some ideas for the syntax are

> * `Parent.field`  // my favorite, as it seems most natural

Reviewer 3: Does this work for tuple structs?

```rust
struct S(u8, u16);
let field = S.0; // yes?
```

Reviewer 2: Yes, this should work with tuple structs, and tuples as well.

`<(u8, u16)>.1 == u16`

Reviewer 4: Why not for `union`? Unlike `enum` there are no representational issues with the discriminant. Every union variant already has a concrete type behaviour is defined only when manually checking the initialized portions anyways so I don't see why it should not be possible. Note that this is vitally important for `MaybeUninit`.

Reviewer 2: Because unions would allow safely going form `&union` to `&field` using `Project`, but this is unsafe. I don't want to have to make a zoo of traits to turn on or off `Project` so `union`s can be punted to another RFC.

Reviewer 4: You're right, too bad. I would have expected it to be possible with the same `unsafe` operations as raw pointers at least. This could however work if the projection of `&_` is not guaranteed available for all fields, e.g. requires an additional `unsafe` trait.

Reviewer 2: I added `union` support by making it unsafe to construct the field types, but this is tenuous (and may be removed before it is accepted). No other type requires unsafe to construct, they may use privacy to force you to use unsafe functions to construct them. But the actual construction is always considered safe.

`enums` will remain unsupported.

```rust
struct Foo {
    bar: Bar
}

struct Foo.bar; // or some other name mangle that doesn't conflict with any other name

impl Field for Foo.bar { ... }
```

These are the core parts of this proposal. Every other part of this proposal can be postponed or dropped without affecting this feature's core principles.

Using these core parts we can build as a library projections through `Pin<&T>`, `Rc<_>` and more. We can then use this to safely project through smart pointers like so.

```rust
let foo   : Pin<Box<Foo>    = Box::pin(immovable);
let foo   : Pin<&mut Foo>   = foo.as_mut();
let field : Pin<&mut Field> = foo.project(Foo.field);
```
But to do safe pin projections we will need to introduce a marker trait. Adding this trait is up for debate.
```rust
/// The only people who can implement `PinProjectable` are the creator of the parent type
/// This allows people to opt-in to allowing their fields to be pin projectable.
/// The guarantee is that once you create `Pin<P<Parent>>`, all of the same guarantees that
/// apply to `Pin<P<Parent>>` also apply to `Pin<P<Field>>`
/// For all `Parent: Unpin`, these can be auto implemented for all of their fields.

> unsafe trait PinProjectable: Field {}

Reviewer 4: For a type to opt-into this for one of its fields this needs to be involve that base type.

```
struct Foo {
   bar: Bar,
}

// The type `Foo.bar` is not defined here but somewhere else (in he compiler).
// Thus can't simply implement a trait for it.
// unsafe impl PinProjectable for Foo.bar { }
```

But an alternative formulation could remedy this, I think:

```
unsafe trait PinProjectable<Parent: ?Sized>: Field<Parent=Parent> { }

unsafe impl PinProjectable<Foo> for Foo.bar { }
```

Reviewer 2: Because `Parent` is an associated type of `Field`, we don't need a parameter on `PinProjectable`. If you want to constrain `Parent` via `PinProjectable`, you can do `PinProjectable<Parent = ...>`

We could say that for all intents and purposes, these field types are defined right next to the parent types. This way you can just implement traits for them.

Reviewer 4: Don't trait coherence rules forbid you from manually implementing the trait if the Parent type does not occur explicitely in the trait definition? Afterall, that would be implementing a third-party trait for a third-party type. Or should the *type* of `Foo.bar` be treated as if occurring in the crate where `Foo` is defined?

Reviewer 2: The idea is that `Foo.bar` would *not* be a third part type to the person who defined `Foo`.

So, this exactly
> Or should the type of Foo.bar be treated as if occurring in the crate where Foo is defined?

This feature will also allow for safer patterns in `unsafe` code that deals with intrusive data strutures via `inverse_*` projections. It will also provide projections through raw pointers, which are currently not possible to do safely without a `#[repr(...)]` attribute (and even then it is easy to make a mistake). This will make `unsafe` code easier to audit and easier to write sound `unsafe` code.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This RFC introduces the idea of field types, and build projections on top of field types.

## Terminology, types, and traits

### Field Types

A field type is a compiler generated type that represents a field. For example,

```rust
struct Person {
    name: String,
    age: u32,
}
```

A field type for the `name` field of `Person` would be `Person.name`.

<sub>Note about syntax: the syntax `Type.field` is going to be used as a placeholder syntax for the rest of this RFC, but this is *not* the final syntax for field types, that decision will be made after this RFC gets accepted but before it gets stabilized.</sub>

Field types serve a few important purposes

> * They are integrated with the new `Project` trait (explained later)

Reviewer 2: Thanks for the catch!

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This section will go in detail about the semantics and implementation of the types, traits, and methods introduced by this RFC.

## Field types

Field types are just sugar for unit types (`struct Foo;`) that are declared right next to the `Parent` type.

Like so,
```rust
struct Person {
    pub name: String,
    pub(super) age: u32,
    id: u32,
}

pub struct Person.name;
pub(super) struct Person.age;
struct Person.id;
```

All field types will implement the following traits: `Debug`, `Clone`, `Copy`, `PartialEq`, `Eq`, `PartialOrd`, `Ord`, `Hash`, with the same implementation for unit types.

They will also need to interact with the privacy rules for fields, and will have the same privacy as the field that they are derived from.

You can make a field type for the following types of types (some example syntax is shown)
* tuples `(T, U, ...)`
    * `<(T, U)>.0`
* tuple structs `Foo(T, U, ...)`
    * `Foo.0`
* structs `struct Foo { field: Field, ... }`
    * `Foo.field`
* unions `union Foo { field: Field }`
    * same syntax as structs

>     * constructing a field type is `unsafe`, this is because accessing fields of `union`s is unsafe

Reviewer 2: Wording, when I wrote his I saw `Parent.field` as a single item, but if we want to see it as two items, `Parent` and `field` then, that's also fine. I think that your suggestion makes the parallel to unions more clear though, so I will keep it.

The compiler can decide whether to actual generate a field type, this is to help compile times (if you don't use field types, then the compiler shouldn't slow down too much because of this feature).

## `*const T`/`*mut T`

> `project_unchecked` and `wrapping_project` will both live inside of `core::ptr`.

Reviewer 2: Ok, that sounds good. I think I should make it clear that these are inherent methods.

We need both `project_unchecked` and `wrapping_project` because there are some important optimization available inside of LLVM related to aliasing and escape analysis. In particular the LLVM `inbounds` assertion tells LLVM that a pointer offset stays within the same allocation and if the pointer is invalid or the offset does not stay within the same allocation it is considered UB. This behaviour is exposed via `project_unchecked`. This can be used by implementations of the `Project` trait for smart pointers that are always valid, like `&T` to enable better codegen. `wrapping_project` on the other hand will not assert `inbounds`, and will just wrap around if the pointer offset is larger than `usize::max_value()`. This safe defined behaviour, even if it is almost always a bug, unlike `project_unchecked` which is UB on invalid pointers or offsets.

> This corresponds with `core::ptr::add` and `core::ptr::wrapping_add` in safety and behaviour. `project_unchecked` and `project` need to be implemented as intrinsics because there is no way to assert that the pointer metadata for fat pointers of `Field::Parent` and `Field::Type` will always match in general without some other compiler support. This is necessary to allow unsized types to be used transparently with this scheme.

Reviewer: So wait, how does this interact with unsized types exactly? This needs to be spelled out.

Since this is specifically to give vocabulary to the direct manipulation of the pointers, I suspect that this should require that the field pointer's metadata is derivable from the parent metadata. Note that this doesn't require equivalent metadata: a theoretical pointer with two metadatas could be split into two child fat pointers with one or the other metadata.

The intrinsics live in [`core::intrinsics`](https://doc.rust-lang.org/core/intrinsics/index.html), of course.

Reviewer 2: > Note that this doesn't require equivalent metadata: a theoretical pointer with two metadatas could be split into two child fat pointers with one or the other metadata.

Actually, you can only have 1 unsized field on a type, and it must be the last field, so the metadata will be exactly the same.

> The intrinsics live in core::intrinsics, of course.

Yes, I will make it clear where everything lives in the next revision. I see now that it wasn't clear enough

> This corresponds with `core::ptr::add` and `core::ptr::wrapping_add` in safety and behaviour. `project_unchecked` and `project` need to be implemented as intrinsics because there is no way to assert that the pointer metadata for fat pointers of `Field::Parent` and `Field::Type` will always match in general without some other compiler support. This is necessary to allow unsized types to be used transparently with this scheme.

Reviewer 2: Same as above, I'll make that clear, I seemed to have overlooked that when I wrote this.

> `inverse_project_unchecked` and `inverse_wrapping_project` are just like their counterparts in safety. `inverse_project_unchecked` is UB to use on invalid pointers, where `inverse_wrapping_project` just wraps around on overflow. But there is one important safety check that must be performed before dereferencing the resulting pointer. The resulting pointer may not actually point to a valid `*const F::Parent` if `*const F::Type` does not live inside of a `F::Parent`, so one must take care to ensure that the parent pointer is indeed valid. This is different from `project_unchecked` and `wrapping_project` because there one only needs to validate the original pointer, not the resulting pointer.

Reviewer: I don't think this is quite right. I think better wording is along the lines of

> `inverse_project_unchecked` is not UB and `inverse_wrapping_project` produces a valid pointer if and only if the initial pointer is both valid and points to a field in the parent type used to perform the inverse projection.

This is all prerequisites on the input pointer to match the field descriptor provided, not something that should (or even could) be checked on the resulting pointer.

This, of course, makes the resulting pointer a lot harder to use safely, as you need more information from outside the type system to make use sound.

Reviewer 2: That wording is much better, thanks! That does look right.

We will add the following methods to raw pointers

```rust
// details about why we need both and what they do exactly in Reference-level explanation

impl<T: ?Sized> *const T {
    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type;
    
    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *const F::Type;

    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
    
    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
}

impl<T: ?Sized> *mut T {
    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
    
    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
    
    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
    
    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
}
```

These will allowing projections through raw pointers without dereferencing the raw pointer. This is useful for building projections through other abstractions like smart pointers (`Rc<T>`, `Pin<&T>`)!

> This is the extent of the core api of this RFC.

Reviewer: Is this line necessary anymore?

Reviewer 2: No, I seemed to have missed that in my last read through last night.

Using this we can do something like this
```rust
struct Foo {
    bar: Bar,
    age: u32
}

struct Bar {
    name: String,
    id: i32
}

let x : Foo = ...;
let y : *const Foo = &x;

let y_bar_name: *const String = unsafe { y.project_unchecked(Foo.bar).project_unchecked(Bar.name) };
```

In the end `y_bar_name` will contain a pointer to `x.bar.name`, all without dereferencing a single pointer! (Given that this is a verbose, we may want some syntax for this, but that is out of scope for this RFC)

> But, we can build on this foundation and create a more power abstraction, to generalize this project notion to smart pointers.

Reviewer: This seems to be out of place now, as the more powerful abstraction isn't provided by this RFC anymore.

Reviewer 2: True, but I still think that it is worth pointing out. Also, I just moved `Project` to the future possibilities section so people can still so what a project trait could look like, and how it would work with this proposal.

We need both `project_unchecked` and `wrapping_project` because there are some important optimization available inside of LLVM related to aliasing and escape analysis. In particular the LLVM `inbounds` assertion tells LLVM that a pointer offset stays within the same allocation and if the pointer is invalid or the offset does not stay within the same allocation it is considered UB. This behaviour is exposed via `project_unchecked`. This can be used by implementations of the `Project` trait for smart pointers that are always valid, like `&T` to enable better codegen. `wrapping_project` on the other hand will not assert `inbounds`, and will just wrap around if the pointer offset is larger than `usize::max_value()`. This safe defined behaviour, even if it is almost always a bug, unlike `project_unchecked` which is UB on invalid pointers or offsets.

> This corresponds with `core::ptr::add` and `core::ptr::wrapping_add` in safety and behaviour.

Reviewer 2: How did I miss that!

`inverse_project_unchecked` and `inverse_wrapping_project` are have all of the same safety requirements as their counterparts, and some more.

> `inverse_project_unchecked` and `inverse_wrapping_project` produces a valid pointer without UB if and only if the initial pointer is both valid and points to a field in the parent type used to perform the inverse projection. `inverse_wrapping_project` will never cause UB, but may produce invalid pointers. `inverse_project_unchecked` will cause UB if the condition above is not met. This is different from `project_unchecked` and `wrapping_project` because they only need to validate the original pointer, not the resulting pointer.

Reviewer: ```suggestion
`inverse_project_unchecked` and `inverse_wrapping_project` produces a valid pointer without UB if and only if the initial pointer is both valid and points to a field in the parent type used to perform the inverse projection. `inverse_wrapping_project` will never cause UB, but may produce invalid pointers. `inverse_project_unchecked` will cause UB if the condition above is not met. This is different from `project_unchecked` and `wrapping_project` because they only need to prove the validity of the original pointer, and not the context in which it resides.
```

Reviewer 2: I think I will keep
> , and not the context in which it resides
but validate means the same thing as
> prove the validity of
so I won't keep that

> `inverse_project_unchecked` and `inverse_wrapping_project` correspond to `core::ptr::sub` and `core::ptr::wrapping_sub` in safety and behaviour.

Reviewer: `<*const _>::sub` and `<*const _>::wrapping_sub`.

For example of where `project_unchecked` would be UB.

```rust
struct Foo {
    bar: Bar
}

let x : *const Foo = 2usize as *const Foo;
let y : *const Bar = x.project_unchecked(Foo.bar); // UB, x does not point to a valid instance of Foo
```

With `wrapping_project`

```rust

> let z : *const Bar = x.wrapping_project(Foo.bar); // not UB, but is still invalid

Reviewer: ```suggestion
let z : *const Bar = x.wrapping_project(Foo.bar); // not UB, but still produces an invalid pointer
```

    - This would require Generic Associated Types at the very least, and maybe some other features like assocaited traits

# Prior art
[prior-art]: #prior-art

- C++'s pointer to members `Parent::*field`
- Java's `class Field`
    - Similar reflection capabilies in other languages

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Syntax for the type fields
    - not to be decided before accepting this RFC, but must be decided before stabilization
    - Some other variations of the syntax are ...
        - `Type::field` // This is bad because it conflicts with associated method, and there isn't a way to disambiguate them easily
        - `Type~field`  // This adds a new sigil to the language
- How will pointer metadata be handled for more exotic Custom DSTs?
    - see @CAD97's example [here](https://github.com/rust-lang/rfcs/pull/2708#discussion_r296933269)
        > Note that this doesn't require equivalent metadata: a theoretical pointer with two metadatas could be split into two child fat pointers with one or the other metadata.

# Future possibilities
[future-possibilities]: #future-possibilities

> - Extend the `Project` trait to implement all smart pointers in the standard library

Reviewer: ```suggestion
```

Trait is now introduced below.
